datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -448,7 +474,21 @@ impl ExpressionVisitor for ExprIdentifierVisitor<'_> {
 
         self.id_array[idx] = (self.series_number, desc.clone());
         self.visit_stack.push(VisitRecord::ExprItem(desc.clone()));
-        let data_type = self.data_type.clone();
+
+        let data_type = if let Ok(data_type) = expr.get_type(&self.input_schema) {
+            data_type
+        } else {
+            // expression type could not be resolved in schema, fall back to all schemas
+            let merged_schema =
+                self.all_schemas
+                    .iter()
+                    .fold(DFSchema::empty(), |mut lhs, rhs| {
+                        lhs.merge(rhs);
+                        lhs
+                    });
```

Comment:
I wonder if we could use `reduce` here (I realize you just refactored this code)
```suggestion
            let merged_schema =
                self.all_schemas
                    .iter()
                    .reduce(|mut lhs, rhs| {
                        lhs.merge(rhs);
                        lhs
                    });
```

Reply:
This doesn't work so well because we iterate over `Arc<DFSchema>` but `merge` has the signature `fn merge(&mut self, other_schema: &DFSchema)`. We need a df schema that we can mutate to begin with, so reducing won't work easily unless we transform the iterator's item type to `&mut DFSchema` first, which likely means a lot of cloning.

I`d leave this code as-is, personally, since it's likely going to be removed anyway.

Reply:
agree